uncertainty_baselines/models/radial_utils.py

```diff
@@ -0,0 +1,447 @@
+    """
+    # x_i = Sum of x_j ** 2 for j = N - i
+    base = tf.reverse(x_c, [-1])
+    base = tf.math.cumsum(base ** 2, -1)
+
+    # Now we flip back which gives us
+    # [x_n**2 + x_{n-1}**2 + ... + x_1**2, ..., x_n**2]
+    base = tf.reverse(base, [-1])
+    base = tf.math.sqrt(base)
+    theta = tf.math.atan(base[..., 1:] / x_c[..., 0:-1]) + (tf.constant(
+      math.pi) / 2)  # The first element is x_0 / base_1
+
+    # Correcting the last element
+    theta[..., -1] = tf.math.atan(x_c[..., -1] /
+                                  (x_c[..., -2] + tf.math.sqrt(
+                                    (x_c[..., -1] ** 2) + (
+                                        x_c[..., -2] ** 2)))) + (
+                         tf.constant(math.pi) / 2)
+
+    # Now we add the radius to the start,
+    # which is just the first term of the base vector that we haven't used
+    x_r = tf.cat((tf.expand_dims(base[..., 0], -1), theta), 1)
+
+    return x_r
+
+  def _sample_n(self, n, seed=None):
+    raw = samplers.normal(
+      shape=tf.concat([[n], self.batch_shape, self.event_shape], axis=0),
+      seed=seed, dtype=self.dtype)
+    direction = raw / tf.norm(raw, ord=2, axis=-1)[..., tf.newaxis]
+    distance = samplers.normal(
+      shape=tf.concat(
+        [[n], self.batch_shape, [1] * len(self.event_shape)], axis=0),
+      seed=seed, dtype=self.dtype)
+    return self._loc + self.scale * direction * distance
+
+  def entropy(self):
+    warnings.warn(
+      "Entropy is correct only up to a constant, for optimization only."
+    )
+    return -tf.math.log(tf.math.reduce_sum(self.scale))
+
+  def mean(self):
+    shape = tensorshape_util.concatenate(self.batch_shape,
+                                         self.event_shape)
+    has_static_shape = tensorshape_util.is_fully_defined(shape)
+    if not has_static_shape:
+      shape = tf.concat([
+        self.batch_shape_tensor(),
+        self.event_shape_tensor(),
+      ], 0)
+
+    if self.loc is None:
+      return tf.zeros(shape, self.dtype)
+
+    if has_static_shape and shape == self.loc.shape:
+      return tf.identity(self.loc)
+
+    # Add dummy tensor of zeros to broadcast. This is only necessary if
+    # shape != self.loc.shape,
+    # but we could not determine if this is the case.
+    return tf.identity(self.loc) + tf.zeros(shape, self.dtype)
+
+  def kl_divergence(self, other, name='kl_divergence'):
+    if (isinstance(other, distributions.independent.Independent) and
+            isinstance(other.distribution, distributions.Normal)):
+      return kl_radial_normal(p=self, q=other)
+    elif isinstance(other, Radial):
+      return kl_radial_radial(p=self, q=other)
+    else:
+      raise NotImplementedError(f'Other distribution has type {type(other)}.')
+
+
+@kullback_leibler.RegisterKL(
+  Radial, distributions.MultivariateNormalLinearOperator)
+def kl_radial_normal(p, q, name=None):
+  with tf.name_scope(name or 'kl_radial_normal'):
```

Comment:
Should we check that p and q are correct distributions here? Or is that handled elsewhere?

Reply:
The registry guarantees this, and in normal execution it is checked in Radial.kl_divergence